### PR TITLE
Prevent DROP/RENAME TABLE IF EXISTS from silently ignoring views

### PR DESCRIPTION
**Pull request by Riddle4045**
## Description

Fixes: #9129 

## Documentation

(:white_check_mark:) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(:white_check_mark:) No release notes entries required.
( ) Release notes entries required with the following suggested text:

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/DropTableTask.java
Comment (findinpath) [reply core/trino-main/src/main/java/io/trino/execution/DropTableTask.java:75]: @losipiuk , @lukasz-stec  please chime on this topic based on your initial discussion from https://github.com/trinodb/trino/pull/8869#discussion_r688293150

Comment (Riddle4045) [reply core/trino-main/src/main/java/io/trino/execution/DropTableTask.java:75]: @findinpath thanks for the review, agree that table existence check is not relevant when the name points to views/materialized views  during `DROP TABLE` operation

Comment (lukasz-stec) [reply core/trino-main/src/main/java/io/trino/execution/DropTableTask.java:75]: IMO, either way is ok (i.e. depends on `IF EXISTS` semantics) but we should have it consistent across all object types. If we are changing this for `table`, we should also change it for `view` and `materialized view` and not only for drop by for `RENAME` as well.

Comment (Riddle4045) [reply core/trino-main/src/main/java/io/trino/execution/DropTableTask.java:75]: @lukasz-stec  Thanks for the comment, I agree that the behavior needs to be consistent across `RENAME` & `DROP` for table/views/materialized views. I can make it consistent as part of this review.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMaterializedViews.java
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMaterializedViews.java:520]: Please do work with :

```
system.metadata.materialized_views
```
for obtaining the materialized views to drop.

Please do the changes in `TestIcebergMaterializedViews` in a separate commit.

Comment (Riddle4045) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMaterializedViews.java]: @findinpath  this is not needed anymore - it's removed in this review #11626

**Review by findinpath — APPROVED** [review]

**Comment by Riddle4045** [comment]
Waiting for #11626  to merge, this will unblock pipeline tests.

**Comment by findinpath** [comment]
Please squash your commits.

File: core/trino-main/src/main/java/io/trino/execution/DropTableTask.java
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/DropTableTask.java:66]: wrong error code

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/DropTableTask.java:66]: cc @findinpath

Comment (findinpath) [reply core/trino-main/src/main/java/io/trino/execution/DropTableTask.java:66]: @findepi: Are you rather leaning here towards `io.trino.spi.StandardErrorCode#UNSUPPORTED_TABLE_TYPE` ?

Comment (Riddle4045) [reply core/trino-main/src/main/java/io/trino/execution/DropTableTask.java:66]: @findepi  there is a `io.trino.spi.StandardErrorCode#GENERIC_USER_ERROR` , do you think that will suffice?

Comment (findinpath) [reply core/trino-main/src/main/java/io/trino/execution/DropTableTask.java:66]: I went over the entries from `StandardErrorCode` and found  the error code `NOT_SUPPORTED` to fit best to this user error type of exception. Moreover the message of the exception fits also quite well this exception error code.

Comment (findinpath) [reply core/trino-main/src/main/java/io/trino/execution/DropTableTask.java:66]: Disregard please my previous comment.
I was drawn to say that `NOT_SUPPORTED` fits best because of the nature of `DROP` statement can be used for tables/views/MV.
However `NOT_SUPPORTED` is rather used to point out that a specific connector has or has not support for a specific statement.

**Review by findinpath — APPROVED** [review]

File: core/trino-main/src/test/java/io/trino/execution/TestDropTableTask.java
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/execution/TestDropTableTask.java:103]: This was added in https://github.com/trinodb/trino/pull/8869
so cc @lukasz-stec @sopel39 @raunaqmorarka

Comment (findepi) [reply core/trino-main/src/test/java/io/trino/execution/TestDropTableTask.java:103]: cc @losipiuk @martint @hashhar  too

**Review by findepi — APPROVED** [review]

File: core/trino-main/src/test/java/io/trino/execution/TestDropTableTask.java
Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/execution/TestDropTableTask.java:103]: should we also do the same change for drop/rename view and materialized view to be consistent?

Comment (findinpath) [reply core/trino-main/src/test/java/io/trino/execution/TestDropTableTask.java:103]: @Riddle4045 please do change the tasks indicated by @lukasz-stec  as well.

File: core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java:72]: We probably should stick with `TABLE_NOT_FOUND`

Comment (Riddle4045) [reply core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java:72]: @sopel39  we had a discussion on this here https://github.com/trinodb/trino/pull/11555#discussion_r836132958

File: core/trino-main/src/main/java/io/trino/execution/DropTableTask.java
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/execution/DropTableTask.java]: > we cannot say "MV is a table of a wrong type", since it;s not a table at all

We use `TABLE_NOT_FOUND` for tables, views and MVs. We don't have specific `VIEW_NOT_FOUND` and `MV_NOT_FOUND` errors. I don't think these are needed either.

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/execution/DropTableTask.java]: > To me GENERIC_USER_ERROR sounds better

What's the difference between this case (an object with different type and same name exists) and case where there is no object with given name?

I cannot rationale why we would like to use `GENERIC_USER_ERROR` here rather than `TABLE_NOT_FOUND`? After all, one can say that trying to drop a table that doesn't exist is also `GENERIC_USER_ERROR`

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/DropTableTask.java]: You said that TABLE_NOT_FOUND is used for tables/views/MVs
... so it's not a clean indication that a relation was found, but of a wrong type

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/execution/DropTableTask.java]: Ok

File: core/trino-main/src/test/java/io/trino/execution/TestDropTableTask.java
Comment (Riddle4045) [reply core/trino-main/src/test/java/io/trino/execution/TestDropTableTask.java:103]: @findinpath thanks, I will revise the changes.

File: core/trino-main/src/main/java/io/trino/execution/DropViewTask.java
Comment (findinpath) [inline core/trino-main/src/main/java/io/trino/execution/DropViewTask.java:75]: In the context of connectors using shared metastores (hive, iceberg, delta) we would have this call throwing a bogus exception about saying "cannot query Iceberg/Delta/Hive table".

We actually want to know here whether the table exists and not get actually get extensive information about the table here.

I propose (in a follow-up PR) that we introduce a method `tableExists(session, name)` in `MetadataManager` as well as on `ConnectorMetadata` to  avoid such exceptions.

cc @findepi @ebyhr

File: core/trino-main/src/test/java/io/trino/execution/TestDropTableTask.java
Comment (findinpath) [reply core/trino-main/src/test/java/io/trino/execution/TestDropTableTask.java:103]: Pls do change `SetPropertiesTask` , `RenameViewTask` as well.

**Review by findinpath — CHANGES_REQUESTED** [review]
Pls do change SetPropertiesTask , RenameViewTask as well.

**Comment by Riddle4045** [comment]
> Pls do change SetPropertiesTask , RenameViewTask as well.

@findinpath  If I understand this corrently , `RenameViewTask` and `SetPropertiesTask` already follow the pattern of this PR. That is the `isExist` check does not change the behavior. Let me know if I am missing something? :)

File: core/trino-main/src/main/java/io/trino/execution/DropViewTask.java
Comment (Riddle4045) [reply core/trino-main/src/main/java/io/trino/execution/DropViewTask.java:75]: @findinpath  good catch, since `getTableHandle` returns optional,  instead of throwing if we defaulted returned an `Opional.OfEmtpy()`  is that behavior in-line with this function call?

Comment (findinpath) [reply core/trino-main/src/main/java/io/trino/execution/DropViewTask.java:75]: @Riddle4045 in the context of using a shared metastore (e.g. : say hive and iceberg use the same Hive Metastore Service) we could have the hive table `foobar` visible in both Hive & Iceberg. However if we are in the context of Iceberg (by previously using `USE iceberg.default`) and trying to drop a view in the following manner `DROP VIEW foobar` (note the unqualified name for the table) we'll get an exception saying `Cannot query Iceberg/Delta/Hive table` - which is a rather misleading  error message.

**Comment by findinpath** [comment]
You have checkstyle violations

```
Error:  src/main/java/io/trino/execution/DropViewTask.java:[34,15] (imports) UnusedImports: Unused import - io.trino.spi.StandardErrorCode.TABLE_NOT_FOUND.
Error:  src/main/java/io/trino/execution/DropMaterializedViewTask.java:[34,15] (imports) UnusedImports: Unused import - io.trino.spi.StandardErrorCode.TABLE_NOT_FOUND.
Error:  src/test/java/io/trino/execution/TestDropViewTask.java:[61] (regexp) RegexpMultiline: Whitespace character before closing parenthesis
Error:  src/test/java/io/trino/execution/TestDropViewTask.java:[61,93] (whitespace) ParenPad: ')' is preceded with whitespace.
Error:  src/test/java/io/trino/execution/TestDropViewTask.java:[61,93] (whitespace) SingleSpaceSeparator: Use a single space to separate non-whitespace characters.
```

File: core/trino-main/src/main/java/io/trino/execution/DropViewTask.java
Comment (Riddle4045) [reply core/trino-main/src/main/java/io/trino/execution/DropViewTask.java:75]: @findinpath  Thanks for the explanation! Makes sense to me now. I'll create an issue for this :).

**Comment by Riddle4045** [comment]
> You have checkstyle violations
> 
> ```
> Error:  src/main/java/io/trino/execution/DropViewTask.java:[34,15] (imports) UnusedImports: Unused import - io.trino.spi.StandardErrorCode.TABLE_NOT_FOUND.
> Error:  src/main/java/io/trino/execution/DropMaterializedViewTask.java:[34,15] (imports) UnusedImports: Unused import - io.trino.spi.StandardErrorCode.TABLE_NOT_FOUND.
> Error:  src/test/java/io/trino/execution/TestDropViewTask.java:[61] (regexp) RegexpMultiline: Whitespace character before closing parenthesis
> Error:  src/test/java/io/trino/execution/TestDropViewTask.java:[61,93] (whitespace) ParenPad: ')' is preceded with whitespace.
> Error:  src/test/java/io/trino/execution/TestDropViewTask.java:[61,93] (whitespace) SingleSpaceSeparator: Use a single space to separate non-whitespace characters.
> ```
> 
> Run locally to make sure that your changes are correct:
> 
> ```
> mvn install -DskipTests -nsu -pl '!:trino-main'
> mvn test-compile -Perrorprone-compiler -pl :trino-main
> ```

yes, sorry I am intransit - pushed a bad change. Let me fix these and start a new iteration.

**Comment by findinpath** [comment]
There's also another failure

```
Error:  io.trino.plugin.hive.TestHiveConnectorTest.testShowCreateView  Time elapsed: 0.632 s  <<< FAILURE!
java.lang.RuntimeException: line 1:1: View 'hive.tpch.test_show_create_view1jvibgexnb' does not exist
	at io.trino.testing.AbstractTestingTrinoClient.execute(AbstractTestingTrinoClient.java:122)

...
Caused by: io.trino.spi.TrinoException: line 1:1: View 'hive.tpch.test_show_create_view1jvibgexnb' does not exist
	at io.trino.sql.analyzer.SemanticExceptions.semanticException(SemanticExceptions.java:48)
	at io.trino.sql.analyzer.SemanticExceptions.semanticException(SemanticExceptions.java:43)
```

This one seems to be an existing bug in Trino that you uncovered with your changes.

**Comment by Riddle4045** [comment]
> There's also another failure
> 
> ```
> Error:  io.trino.plugin.hive.TestHiveConnectorTest.testShowCreateView  Time elapsed: 0.632 s  <<< FAILURE!
> java.lang.RuntimeException: line 1:1: View 'hive.tpch.test_show_create_view1jvibgexnb' does not exist
> 	at io.trino.testing.AbstractTestingTrinoClient.execute(AbstractTestingTrinoClient.java:122)
> 
> ...
> Caused by: io.trino.spi.TrinoException: line 1:1: View 'hive.tpch.test_show_create_view1jvibgexnb' does not exist
> 	at io.trino.sql.analyzer.SemanticExceptions.semanticException(SemanticExceptions.java:48)
> 	at io.trino.sql.analyzer.SemanticExceptions.semanticException(SemanticExceptions.java:43)
> ```
> 
> This one seems to be an existing bug in Trino that you uncovered with your changes.
> 
> This one can be tackled by changing `DropViewTask` to do the handling in the following sequence:
> 
> ```
> check if materialized view and throw exception
> check if view and do happy case handling
> check if table and throw exception
> if statement is not an `IF EXISTS`  statement throw exception
> ```

@findinpath  Current order of the checks seems correct to me

```
1. check if materialized views -  if yes throw exception
2. check if not a view - f yes check if it's table - if yes throw specific exception. if no - throw generic exception.
3. check access control 
4. perform operation

```

I think the issue here is that the Hive metadata provider returns `false` for `isView` check in the case of non-existent views.  Which is why we get the generic exception from `step 2`. We should also probably catch the exceptions in `step 4` in case DDL has an `isExists`

what do you think?

**Comment by findinpath** [comment]
@Riddle4045  this is what i had in mind for `DropViewTask`
(this code snippet should be considered as a draft and not a final version for this PR)

```
    @Override
    public ListenableFuture<Void> execute(
            DropView statement,
            QueryStateMachine stateMachine,
            List<Expression> parameters,
            WarningCollector warningCollector)
    {
        Session session = stateMachine.getSession();
        QualifiedObjectName name = createQualifiedObjectName(session, statement, statement.getName());

        if (metadata.isMaterializedView(session, name)) {
            throw semanticException(
                    GENERIC_USER_ERROR,
                    statement,
                    "View '%s' does not exist, but a materialized view with that name exists. Did you mean DROP MATERIALIZED VIEW %s?", name, name);
        }

        if (metadata.isView(session, name)) {
            accessControl.checkCanDropView(session.toSecurityContext(), name);

            metadata.dropView(session, name);

            return immediateVoidFuture();
        }

        // This can only be a table
        Optional<TableHandle> table = metadata.getTableHandle(session, name);
        if (table.isPresent()) {
            throw semanticException(
                    GENERIC_USER_ERROR,
                    statement,
                    "View '%s' does not exist, but a table with that name exists. Did you mean DROP TABLE %s?", name, name);
        }
        if (!statement.isExists()) {
            throw semanticException(GENERIC_USER_ERROR, statement, "View '%s' does not exist", name);
        }
        return immediateVoidFuture();
    }
```

**Comment by Riddle4045** [comment]
> 

@findinpath  Great! Should I extend the behavior to all the other tasks  ( DropTable, Materializedview etc ) as well?

**Comment by findinpath** [comment]
@Riddle4045  please check the build failures.

**Comment by Riddle4045** [comment]
> 

@findinpath  done.

**Comment by Riddle4045** [comment]
> > 
> 
> @findinpath done. Also created a new issue for our discussion on cases with shared metastore earlier. #12069

@findinpath  I don't have permissions to merge this request, let me know what I need to do in order to merge this :) thanks!

**Review by findinpath — APPROVED** [review]

**Review by findepi — APPROVED** [review]

**Comment by Riddle4045** [comment]
@findinpath  can you approve again, thanks!

**Pull request merged**
